File: sdk/typescript/src/utils/requestDeduplicator.ts
// SPDX-License-Identifier: Apache-2.0
// Copyright (c) 2026 KirkyX. All rights reserved.

/**
 * Request deduplication utility for SDK operations.
 *
 * This module provides functionality to prevent duplicate requests from being sent
 * to the server within a short time window. This is useful for:
 * - Reducing server load
 * - Improving response speed (return cached result immediately)
 * - Preventing duplicate operations (e.g., duplicate registration)
 */

export interface PendingRequest {
  timestamp: number;
  promise: Promise<any>;
}

export interface DeduplicatorStats {
  hits: number;
  misses: number;
  errors: number;
  hitRate: number;
  pendingCount: number;
  completedCount: number;
  ttlSeconds: number;
}

/**
 * Prevents duplicate requests from being sent within a short time window.
 *
 * This class tracks in-flight requests and returns the same result to all
 * concurrent requests for the same endpoint and parameters.
 *
 * @example
 * ```typescript
 * const deduplicator = new RequestDeduplicator({ ttlSeconds: 5.0 });
 *
 * // First request
 * const result1 = await deduplicator.execute(
 *   'register_public_key',
 *   { channel_id: 'test', public_key: '...' },
 *   () => apiCall()
 * );
 *
 * // Concurrent duplicate request (will wait for first request)
 * const result2 = await deduplicator.execute(
 *   'register_public_key',
 *   { channel_id: 'test', public_key: '...' },
 *   () => apiCall()
 * );
 * // result1 === result2 (same object)
 * ```
 */
export class RequestDeduplicator {
  private pending: Map<string, PendingRequest>;
  private completed: Map<string, any>;
  private hits: number = 0;
  private misses: number = 0;
  private errors: number = 0;

  constructor(
    private options: {
      ttlSeconds?: number;
      maxPending?: number;
      maxCached?: number;
    } = {}
  ) {
    const {
      ttlSeconds = 5.0,
      maxPending = 1000,
      maxCached = 10000,
    } = options;

    this.options.ttlSeconds = ttlSeconds;
    this.options.maxPending = maxPending;
    this.options.maxCached = maxCached;

    this.pending = new Map();
    this.completed = new Map();
  }

  /**
   * Generate a unique key for the request.
   */
  private generateKey(endpoint: string, params: Record<string, any> | undefined): string {
    // Create a deterministic string from the parameters
    const paramsStr = params ? JSON.stringify(params, Object.keys(params).sort()) : '';
    const key = `${endpoint}:${paramsStr}`;

    // Use SHA256 for better distribution
    let hash = 0;
    for (let i = 0; i < key.length; i++) {
      const char = key.charCodeAt(i);
      hash = ((hash << 5) - hash) + char;
      hash = hash & hash; // Convert to 32bit integer
    }
    return `req_${hash}`;
  }

  /**
   * Execute a request with deduplication.
   *
   * @param endpoint - API endpoint
   * @param params - Request parameters
   * @param func - Async function to execute the request
   * @param useCache - Whether to use completed request cache
   * @returns Result from the request function
   */
  async execute<T>(
    endpoint: string,
    params: Record<string, any> | undefined,
    func: () => Promise<T>,
    useCache: boolean = true
  ): Promise<T> {
    const key = this.generateKey(endpoint, params);

    // Check completed cache first
    if (useCache && this.completed.has(key)) {
      this.hits++;
      return this.completed.get(key);
    }

    // Check pending requests
    if (this.pending.has(key)) {
      // Request is pending, wait for result
      this.hits++;
      const pending = this.pending.get(key)!;
      return pending.promise as T;
    }

    // Execute the request
    this.misses++;

    // Create a promise for this request
    const promise = (async () => {
      try {
        const result = await func();

        // Store result in completed cache
        if (useCache) {
          if (this.completed.size >= (this.options.maxCached || 10000)) {
            // Remove oldest entry (simple FIFO)
            const firstKey = this.completed.keys().next().value;
            this.completed.delete(firstKey);
          }
          this.completed.set(key, result);
        }

        return result;
      } catch (error) {
        this.errors++;
        throw error;
      } finally {
        // Remove from pending
        this.pending.delete(key);
      }
    })();

    // Store pending request
    if (this.pending.size >= (this.options.maxPending || 1000)) {
      // Remove oldest pending request
      const firstKey = this.pending.keys().next().value;
      this.pending.delete(firstKey);
    }

    this.pending.set(key, {
      timestamp: Date.now(),
      promise,
    });

    return promise;
  }

  /**
   * Remove expired entries from completed cache.
   *
   * @returns Number of entries removed
   */
  cleanupExpired(): number {
    let removed = 0;
    const cutoff = Date.now() - ((this.options.ttlSeconds || 5.0) * 1000);

    // Remove oldest entries if we exceed maxCached
    while (this.completed.size > (this.options.maxCached || 10000) * 2) {
      const firstKey = this.completed.keys().next().value;
      if (firstKey) {
        this.completed.delete(firstKey);
        removed++;
      }
    }

    return removed;
  }

  /**
   * Clear all pending requests.
   *
   * @returns Number of pending requests cleared
   */
  clearPending(): number {
    const count = this.pending.size;
    this.pending.clear();
    return count;
  }

  /**
   * Clear all completed requests.
   *
   * @returns Number of completed requests cleared
   */
  clearCompleted(): number {
    const count = this.completed.size;
    this.completed.clear();
    return count;
  }

  /**
   * Clear all pending and completed requests.
   *
   * @returns Total number of requests cleared
   */
  clearAll(): number {
    const pendingCount = this.clearPending();
    const completedCount = this.clearCompleted();
    return pendingCount + completedCount;
  }

  /**
   * Get statistics about the deduplicator.
   *
   * @returns Dictionary with statistics
   */
  getStats(): DeduplicatorStats {
    const totalRequests = this.hits + this.misses;
    const hitRate = totalRequests > 0 ? this.hits / totalRequests : 0.0;

    return {
      hits: this.hits,
      misses: this.misses,
      errors: this.errors,
      hitRate,
      pendingCount: this.pending.size,
      completedCount: this.completed.size,
      ttlSeconds: this.options.ttlSeconds || 5.0,
    };
  }

  /**
   * Reset statistics counters.
   */
  resetStats(): void {
    this.hits = 0;
    this.misses = 0;
    this.errors = 0;
  }
}

File: sdk/rust/src/utils/request_deduplicator.rs
// SPDX-License-Identifier: Apache-2.0
// Copyright (c) 2026 KirkyX. All rights reserved.

use std::collections::HashMap;
use std::sync::Arc;
use std::time::{Duration, Instant};
use sha2::{Sha256, Digest};
use serde::{Serialize, Deserialize};
use tokio::sync::Mutex as TokioMutex;

#[derive(Debug)]
struct PendingRequest {
    _timestamp: Instant,
    result: Option<Result<String, String>>,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct DeduplicatorStats {
    pub hits: u64,
    pub misses: u64,
    pub errors: u64,
    pub hit_rate: f64,
    pub pending_count: usize,
    pub completed_count: usize,
    pub ttl_seconds: f64,
}

pub struct RequestDeduplicator {
    pending: Arc<TokioMutex<HashMap<String, PendingRequest>>>,
    completed: Arc<TokioMutex<HashMap<String, String>>>,
    _ttl: Duration,
    max_pending: usize,
    max_completed: usize,
    stats: Arc<TokioMutex<DeduplicatorStats>>,
}

impl RequestDeduplicator {
    pub fn new(ttl_seconds: f64, max_pending: usize, max_completed: usize) -> Self {
        Self {
            pending: Arc::new(TokioMutex::new(HashMap::new())),
            completed: Arc::new(TokioMutex::new(HashMap::new())),
            _ttl: Duration::from_secs_f64(ttl_seconds),
            max_pending,
            max_completed,
            stats: Arc::new(TokioMutex::new(DeduplicatorStats {
                hits: 0,
                misses: 0,
                errors: 0,
                hit_rate: 0.0,
                pending_count: 0,
                completed_count: 0,
                ttl_seconds,
            })),
        }
    }

    pub fn default() -> Self {
        Self::new(5.0, 1000, 10000)
    }

    fn generate_key(&self, endpoint: &str, params: &Option<serde_json::Value>) -> String {
        let params_str = if let Some(p) = params {
            serde_json::to_string(p).unwrap_or_default()
        } else {
            String::new()
        };

        let key = format!("{}:{}", endpoint, params_str);

        // Use SHA256 for better distribution
        let mut hasher = Sha256::new();
        hasher.update(key.as_bytes());
        let result = hasher.finalize();

        format!("req_{:x}", result)
    }

    pub async fn execute<F, Fut>(
        &self,
        endpoint: &str,
        params: Option<serde_json::Value>,
        func: F,
        use_cache: bool,
    ) -> Result<String, String>
    where
        F: FnOnce() -> Fut + Send,
        Fut: std::future::Future<Output = Result<String, String>> + Send,
    {
        let key = self.generate_key(endpoint, &params);

        if use_cache {
            let completed = self.completed.lock().await;
            if let Some(result) = completed.get(&key) {
                let mut stats = self.stats.lock().await;
                stats.hits += 1;
                stats.hit_rate = stats.hits as f64 / (stats.hits + stats.misses) as f64;
                return Ok(result.clone());
            }
        }

        {
            let mut pending = self.pending.lock().await;
            if let Some(_pending_req) = pending.get_mut(&key) {
                {
                    let mut stats = self.stats.lock().await;
                    stats.hits += 1;
                    stats.hit_rate = stats.hits as f64 / (stats.hits + stats.misses) as f64;
                }

                // Simple polling for now (in a real implementation, use a condition variable)
                drop(pending);
                tokio::time::sleep(Duration::from_millis(10)).await;

                let pending = self.pending.lock().await;
                if let Some(pending_req) = pending.get(&key) {
                    if let Some(ref result) = pending_req.result {
                        return result.clone();
                    }
                }
            }
        }

        {
            let mut stats = self.stats.lock().await;
            stats.misses += 1;
        }

        {
            let mut pending = self.pending.lock().await;
            if pending.len() >= self.max_pending {
                if let Some(oldest_key) = pending.keys().next().cloned() {
                    pending.remove(&oldest_key);
                }
            }
            pending.insert(key.clone(), PendingRequest {
                _timestamp: Instant::now(),
                result: None,
            });
        }

        let result = func().await;

        {
            let mut pending = self.pending.lock().await;
            if let Some(pending_req) = pending.get_mut(&key) {
                pending_req.result = Some(result.clone());
            }
            pending.remove(&key);
        }

        if use_cache && result.is_ok() {
            let mut completed = self.completed.lock().await;
            if completed.len() >= self.max_completed {
                // Remove oldest entry (simple FIFO)
                if let Some(oldest_key) = completed.keys().next().cloned() {
                    completed.remove(&oldest_key);
                }
            }
            if let Ok(ref value) = result {
                completed.insert(key, value.clone());
            }
        }

        {
            let mut stats = self.stats.lock().await;
            if result.is_err() {
                stats.errors += 1;
            }
        }

        result
    }

    pub async fn cleanup_expired(&self) -> usize {
        let mut removed = 0;

        // Remove oldest entries if we exceed max_completed
        let mut completed = self.completed.lock().await;
        while completed.len() > self.max_completed * 2 {
            if let Some(oldest_key) = completed.keys().next().cloned() {
                completed.remove(&oldest_key);
                removed += 1;
            }
        }

        removed
    }

    pub async fn clear_pending(&self) -> usize {
        let mut pending = self.pending.lock().await;
        let count = pending.len();
        pending.clear();
        count
    }

    pub async fn clear_completed(&self) -> usize {
        let mut completed = self.completed.lock().await;
        let count = completed.len();
        completed.clear();
        count
    }

    pub async fn clear_all(&self) -> usize {
        self.clear_pending().await + self.clear_completed().await
    }

    pub async fn get_stats(&self) -> DeduplicatorStats {
        let stats = self.stats.lock().await;
        let pending = self.pending.lock().await;
        let completed = self.completed.lock().await;

        DeduplicatorStats {
            hits: stats.hits,
            misses: stats.misses,
            errors: stats.errors,
            hit_rate: stats.hit_rate,
            pending_count: pending.len(),
            completed_count: completed.len(),
            ttl_seconds: stats.ttl_seconds,
        }
    }

    pub async fn reset_stats(&self) {
        let mut stats = self.stats.lock().await;
        stats.hits = 0;
        stats.misses = 0;
        stats.errors = 0;
        stats.hit_rate = 0.0;
    }
}